### PR TITLE
chore(codegen): respect GOFLAGS when compiling code-generator

### PR DESCRIPTION
**Pull request by yue9944882**
Respect `GOFLAGS` env var which is inherited from Makefile.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by yue9944882** [comment]
/assign @deads2k 

ptal 😃

**Comment by yue9944882** [comment]
/test pull-kubernetes-integration

small world. rs flakiness again :P

**Comment by deads2k** [comment]
@p0lyn0mial you were in the code generator recently.  Can you have a look?

/assign @p0lyn0mial

File: staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh [hunk 1]
```diff
@@ -60,7 +60,7 @@ ${CLIENTGEN} --clientset-name="clientset" ${INPUT_BASE} --input apiregistration/
 
 echo "Building lister-gen"
 listergen="${PWD}/lister-gen"
-go build -o "${listergen}" ${CODEGEN_PKG}/cmd/lister-gen
+go build ${GOFLAGS:-} -o "${listergen}" ${CODEGEN_PKG}/cmd/lister-gen
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh]: this lgtm, but I don't know why we need to pass those flags anyway ?

I have noticed that some `sub-projects` call directy `code-gen/generate-groups.sh` and `code-gen/generate-internal-groups.sh` scripts. For example https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/metrics/hack/update-codegen.sh

However the scripts rely on `go install` to build binaries
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/generate-groups.sh#L53, so I think that they don't respect `GOFLAGS`

**Comment by p0lyn0mial** [comment]
> i added GOFLAGS into generater-groups.sh

sorry, missed that one :). What about https://github.com/kubernetes/code-generator/blob/master/generate-internal-groups.sh script ? Should it also respect `GOFLAGS` ?

**Comment by p0lyn0mial** [comment]
Do you think it would be worth refactoring this script to make use of `generate-groups.sh` ?

Just like the following script relies on it https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/metrics/hack/update-codegen.sh

**Comment by yue9944882** [comment]
> Do you think it would be worth refactoring this script to make use of generate-groups.sh ?

sgtm @p0lyn0mial. ptal the latest change🙂

**Comment by p0lyn0mial** [comment]
If I'm not mistaken before the script generated `clientset`, `informers` and `listers` for the types defined in `k8s.io/kube-aggregator/pkg/apis`. After the changes it generates `deepcopy` and `conversion` (generate-internal-groups.sh) and `clientset` (generate-groups.sh)

I think that we need to slightly change to the script to preserve existing functionallity.
After updating please run the script and make sure that it produces the same output as in `k8s.io/kube-aggregator/pkg/client`

**Comment by yue9944882** [comment]
> I think that we need to slightly change to the script to preserve existing functionallity.

👍🏻thx. now we just make sure `deepcopy,client,lister,informer` is generated for aggregator. ptal

File: staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh [hunk 2]
```diff
@@ -18,65 +18,25 @@ set -o errexit
...
-  --listers-package k8s.io/kube-aggregator/pkg/client/listers \
-  --output-package k8s.io/kube-aggregator/pkg/client/informers \
+SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+CODEGEN_PKG=${CODEGEN_PKG:-$(cd ${SCRIPT_ROOT}; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
+
+CLIENTSET_NAME_VERSIONED=clientset \
+CLIENTSET_PKG_NAME=clientset_generated \
+CLIENTSET_NAME_INTERNAL=internalclientset \
+CLIENTSET_OUTPUT_PATH="k8s.io/kube-aggregator/pkg/client/clientset_generated" \
+${CODEGEN_PKG}/generate-internal-groups.sh deepcopy,client,lister,informer \
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh]: you may want to pass `conversion` too for the internal types.

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh [hunk 3]
```diff
@@ -85,11 +86,11 @@ if [ "${GENS}" = "all" ] || grep -qw "conversion" <<<"${GENS}"; then
 fi
 
 if [ "${GENS}" = "all" ] || grep -qw "client" <<<"${GENS}"; then
-  echo "Generating clientset for ${GROUPS_WITH_VERSIONS} at ${OUTPUT_PKG}/clientset"
+  echo "Generating clientset for ${GROUPS_WITH_VERSIONS} at ${CLIENTSET_OUTPUT_PATH:-${OUTPUT_PKG}/clientset}"
   if [ -n "${INT_APIS_PKG}" ]; then
-    ${GOPATH}/bin/client-gen --clientset-name ${CLIENTSET_NAME_INTERNAL:-internalversion} --input-base "" --input $(codegen::join , $(printf '%s/ ' "${INT_FQ_APIS[@]}")) --output-package ${OUTPUT_PKG}/clientset "$@"
+    ${GOPATH}/bin/client-gen --clientset-name ${CLIENTSET_NAME_INTERNAL:-internalversion} --input-base "" --input $(codegen::join , $(printf '%s/ ' "${INT_FQ_APIS[@]}")) --output-package ${CLIENTSET_OUTPUT_PATH:-${OUTPUT_PKG}/clientset} "$@"
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/code-generator/generate-internal-groups.sh]: I'm not sure whether we should allow for changing `clientset` directory ? Perhaps it's better if everyone generates into the same directory ? Would be nice to know what @sttts thinks about that ?

Maybe we should leave it as it was and update import paths instead :)

Comment (yue9944882) [reply staging/src/k8s.io/code-generator/generate-internal-groups.sh]: a fan of this idea. @sttts wdyt?

Comment (sttts) [reply staging/src/k8s.io/code-generator/generate-internal-groups.sh]: We have old code, e.g. in apiextensions-apiserver where people depend on the location of the clientset package. That's why this customization was added.

File: staging/src/k8s.io/code-generator/generate-groups.sh [hunk 4]
```diff
@@ -85,7 +86,7 @@ if [ "${GENS}" = "all" ] || grep -qw "informer" <<<"${GENS}"; then
   echo "Generating informers for ${GROUPS_WITH_VERSIONS} at ${OUTPUT_PKG}/informers"
   ${GOPATH}/bin/informer-gen \
            --input-dirs $(codegen::join , "${FQ_APIS[@]}") \
-           --versioned-clientset-package ${OUTPUT_PKG}/clientset/${CLIENTSET_NAME_VERSIONED:-versioned} \
+           --versioned-clientset-package ${OUTPUT_PKG}/${CLIENTSET_PKG_NAME:-clientset}/${CLIENTSET_NAME_VERSIONED:-versioned} \
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/generate-groups.sh]: why is this added?

File: staging/src/k8s.io/code-generator/generate-internal-groups.sh [hunk 5]
```diff
@@ -69,6 +69,7 @@ for GVs in ${GROUPS_WITH_VERSIONS}; do
   done
 done
 
+
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/generate-internal-groups.sh]: why this line?

File: staging/src/k8s.io/code-generator/generate-groups.sh [hunk 6]
```diff
@@ -66,14 +66,15 @@ for GVs in ${GROUPS_WITH_VERSIONS}; do
...
 if [ "${GENS}" = "all" ] || grep -qw "deepcopy" <<<"${GENS}"; then
   echo "Generating deepcopy funcs"
   ${GOPATH}/bin/deepcopy-gen --input-dirs $(codegen::join , "${FQ_APIS[@]}") -O zz_generated.deepcopy --bounding-dirs ${APIS_PKG} "$@"
 fi
 
 if [ "${GENS}" = "all" ] || grep -qw "client" <<<"${GENS}"; then
-  echo "Generating clientset for ${GROUPS_WITH_VERSIONS} at ${OUTPUT_PKG}/clientset"
-  ${GOPATH}/bin/client-gen --clientset-name ${CLIENTSET_NAME_VERSIONED:-versioned} --input-base "" --input $(codegen::join , "${FQ_APIS[@]}") --output-package ${OUTPUT_PKG}/clientset "$@"
+  echo "Generating clientset for ${GROUPS_WITH_VERSIONS} at ${CLIENTSET_OUTPUT_PATH:-${OUTPUT_PKG}/clientset}"
+  ${GOPATH}/bin/client-gen --clientset-name ${CLIENTSET_NAME_VERSIONED:-versioned} --input-base "" --input $(codegen::join , "${FQ_APIS[@]}") --output-package ${CLIENTSET_OUTPUT_PATH:-${OUTPUT_PKG}/clientset} "$@"
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/generate-groups.sh]: why do we need CLIENTSET_OUTPUT_PATH ?

File: staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh [hunk 7]
```diff
@@ -18,65 +18,25 @@ set -o errexit
...
-  --versioned-clientset-package k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset \
-  --internal-clientset-package k8s.io/kube-aggregator/pkg/client/clientset_generated/internalclientset \
-  --listers-package k8s.io/kube-aggregator/pkg/client/listers \
-  --output-package k8s.io/kube-aggregator/pkg/client/informers \
+SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+CODEGEN_PKG=${CODEGEN_PKG:-$(cd ${SCRIPT_ROOT}; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
+
+CLIENTSET_NAME_VERSIONED=clientset \
+CLIENTSET_PKG_NAME=clientset_generated \
+CLIENTSET_NAME_INTERNAL=internalclientset \
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh]: that would be awesome, @yue9944882 wdyt ?

File: staging/src/k8s.io/code-generator/generate-groups.sh (hunk 4 above)
Comment (yue9944882) [reply staging/src/k8s.io/code-generator/generate-groups.sh]: b/c the generated clientset pkg has a hardcoded "clientset" in its path. but for the aggregator clientset, we have "clientset_generated" instead.

File: staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (yue9944882) [reply staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh]: its pruned now 👍🏻

File: staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh [hunk 8]
```diff
@@ -18,65 +18,25 @@ set -o errexit
...
-  --output-base ${SCRIPT_BASE} \
-  --input-dirs k8s.io/kube-aggregator/pkg/apis/apiregistration --input-dirs k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1,k8s.io/kube-aggregator/pkg/apis/apiregistration/v1 \
-  --versioned-clientset-package k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset \
-  --internal-clientset-package k8s.io/kube-aggregator/pkg/client/clientset_generated/internalclientset \
-  --listers-package k8s.io/kube-aggregator/pkg/client/listers \
-  --output-package k8s.io/kube-aggregator/pkg/client/informers \
+SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+CODEGEN_PKG=${CODEGEN_PKG:-$(cd ${SCRIPT_ROOT}; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
+
+CLIENTSET_NAME_VERSIONED=clientset \
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh]: ic. So we need this.

**Comment by p0lyn0mial** [comment]
@yue9944882 many thanks for letting me know. So what do we do ? 
We are going to revert all the changes except leaving passing `GOFLAGS` ? 
Or we are fine with the changes made to `generate-groups.sh` and `generate-internal-groups.sh` scripts ?

**Comment by yue9944882** [comment]
> Or we are fine with the changes made to generate-groups.sh and generate-internal-groups.sh scripts ?

@p0lyn0mial i'm leaning on keeping it since it's a good change. wdyt?

**Comment by yue9944882** [comment]
sorted into 2 logical commits. kindly ping @sttts
